### DISCUSSION TITLE
Can't read settings .env file

### DISCUSSION THREAD

**Discussion by Master-Y0da** [fastapi/fastapi#8964] [Questions] [answered]
Not working for me...

    db_url="postgres" # .env file
   
    class Settings(BaseSettings):
       db_url: str

       class Config:
          env_file = ".env"

     settings = Settings()

  
     @lru_cache()
     def get_settings():
       return {
        "db_url": settings.db_url,
     }

Getting error -> **pydantic.error_wrappers.ValidationError: 1 validation error for Settings db_url field required (type=value_error.missing)**

**Comment by ycd** [comment]
Do you have `python-dotenv` installed in your current working environment?

Also, try writing `db_url` without quotes, see examples from [`python-dotenv`'s documentation](https://pypi.org/project/python-dotenv/)

**Comment by Master-Y0da** [comment]
@ycd Yes I have...I just follow the example here [docs](https://fastapi.tiangolo.com/advanced/settings/#reading-a-env-file)

**Comment by pcrespov** [comment]
Just to be sure ... , this script does not run in your setup?

```python
from pydantic import BaseSettings 
from functools import lru_cache
import dotenv


with open(".env", "wt") as fh:
    print("db_url=postgres", file=fh)


class Settings(BaseSettings):
   db_url: str

   class Config:
      env_file = ".env"


settings = Settings()

@lru_cache()
def get_settings():
   return {
    "db_url": settings.db_url,
 }

assert get_settings() == {'db_url': 'postgres'}
```

**Comment by rexbrandy** [comment]
I'm having the same issue.
Here is my file structure
```

├── api
│   ├── config.py
│   ├── core
│   ├── crud.py
│   ├── database.py
│   ├── __init__.py
│   ├── main.py
│   ├── models.py
│   ├── schemas.py
│   ├── tests.py
│   └── utils.py
├── api_db.db
├──-.env
```

Here is my `api/config.py` file.
```
from pydantic import BaseModel

class Settings(BaseModel):
    app_name: str
    debug: bool = True

    cors_allowed_origins = ['*']

    smtp_ssl_port: int = 465
    admin_email: str
    admin_email_pass: str

    sqlalchemy_database_url: str = 'sqlite:///./api_db.db'

    class Config:
        env_prefix = ''
        env_file = '.env' 

settings = Settings()
```

Here is the end of the traceback.
```

  File "pydantic/main.py", line 362, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for Settings
app_name
  field required (type=value_error.missing)
admin_email
  field required (type=value_error.missing)
admin_email_pass
  field required (type=value_error.missing)

```

To me it seems this Config class isn't picking up the .env file.
I have all dependencies installed and followed the FastAPI docs, and the pydantic docs too.

**Comment by ycd** [comment]
@rexbrandy try setting your `env_file` like this `env_file = "../.env"`

**Comment by rexbrandy** [comment]
@ycd tried that and to no avail. 
tried moving the `.env` file into the `api/` folder and also not working.
Is there a way to test if the `Config` class is picking up the file??
If i use `load_dotenv()` in the file that returns True, but the Config is still not working.
I've also tried `export APP_NAME='blah'` and `export app_name='blah'` both don't get picked up.
I think this is move of an issue with pydantic than FastAPI.

**Comment by rexbrandy** [comment]
@lvar  Here is my current work around
```
from pydantic import BaseModel
from dotenv import load_dotenv
import os

load_dotenv(verbose=True)

class Settings(BaseModel):
    app_name: str = os.getenv("APP_NAME")
    debug: bool = True

    cors_allowed_origins = ['*']

    smtp_ssl_port: int = 465
    admin_email: str = os.getenv('ADMIN_EMAIL')
    admin_email_pass: str = os.getenv('ADMIN_EMAIL_PASS')

    sqlalchemy_database_url: str = 'sqlite:///./api_db.db'

    class Config:
        env_prefix = ''
        env_file = '../.env' 

settings = Settings()
```

**Comment by Master-Y0da** [comment]
@rexbrandy You found any solution ?...still getting the same error here.
Maybe is better to use env variables with os...

**Comment by Mause** [comment]
@rexbrandy have you tried using `BaseSettings` instead of `BaseModel`? The `BaseModel` doesn't include any environment variable integration

**Comment by Master-Y0da** [comment]
This still does not work for me 🤒

**Comment by pcrespov** [comment]
@Ivar, did you try adding the suggestion of @rexbrandy: ``load_dotenv(verbose=True)`` ? 

After upgrading requirements in an old project, I suddenly started experiencing this problem. **I still did not investigate** but it might be that at some point ``dotenv``'s load policy has changed.

**Comment by Master-Y0da** [comment]
@pcrespov I already try that...but not working.

**Comment by Mause** [comment]
@lvar perhaps you could provide a self contained reproduction of the issue so we can help further?

**Comment by chrisdietr** [comment]
In https://pypi.org/project/python-dotenv/ you can specifically call the path to your .env file:

```
...
# OR, explicitly providing path to '.env'
from pathlib import Path  # Python 3.6+ only
env_path = Path('.') / '.env'
load_dotenv(dotenv_path=env_path)
```

I just got my .env to work by loading it in the main.py file before initiating the API.

![2020-11-29 at 22 39](https://user-images.githubusercontent.com/2683346/100553033-e0523f80-3293-11eb-8946-17bfccd07d55.png)

**Comment by Mause** [comment]
Pydantic uses that library under the hood anyway, and you can pass a .env file path to the BaseSettings subclass

**Comment by chrisdietr** [comment]
> Pydantic uses that library under the hood anyway, and you can pass a .env file path to the BaseSettings subclass

Thanks I've just learned more about BaseSettings with Pydantic and how to use .env here: https://pydantic-docs.helpmanual.io/usage/settings/#dotenv-env-support - still not sure how to properly access the .env file then if I don't use dotenv syntax, but will figure it out soon.

**Comment by Mause** [comment]
What do you mean by "dotenv syntax"?

**Comment by chrisdietr** [comment]
> What do you mean by "dotenv syntax"?

Hi I have now both versions to working

With Pydantic BaseSetting:
```
# pydantic config file
class DatabaseSettings(BaseSettings):
    DB_URL: str
    DB_NAME: str

    class Config:
        env_file = '.env'
        env_file_encoding = 'utf-8'

# main.py
@app.on_event("startup")
async def startup_db_client():
    app.mongodb_client = AsyncIOMotorClient(settings.DB_URL)
    app.mongodb = app.mongodb_client[settings.DB_NAME]
    
```

With dotenv syntax:
```
# main.py
@app.on_event("startup")
async def startup_db_client():
    # Instead of Pydantic BaseSettings use .env
    app.mongodb_client = AsyncIOMotorClient(os.getenv("DB_URL"))
    app.mongodb = app.mongodb_client[os.getenv("DB_NAME")]
```

All good on my end for now...

**Comment by objectspace** [comment]
`
from pathlib import Path

class Settings(BaseSettings):

    def __init__(self):
        parent_path = Path.cwd()
        super(Settings, self).__init__(
            _env_file=f'{parent_path}/app/env/.env',
            _env_file_encoding='utf-8')
`

This approach works for me.

**Comment by hellocoder2020** [comment]
It depends on where you place your .env file at bro.
If you place .env in root, so just load `env_file = '.env'`
If you place .env nested directory, e.g 

`├── app`
`│   ├── config.py`
`│   ├── .env`

So just `env_file = 'app/.env'`


[Checkout More](https://youtu.be/2DfaD595Wi0)

**Comment by NothingCtrl** [comment]
Try create setting attribute in Settings object, example:

* folder structure
`|-- api`
`|  |-- config.py`
`|  |-- .env`

* `config.py`
```py
class Settings(BaseSettings):
    app_name: str = ""

    class Config:
        env_prefix = ''
        env_file = "api/.env"
        env_file_encoding = 'utf-8'
```

* `.env`
```env
APP_NAME=FastAPI
```

* using
```py
settings = config.Settings()
print(settings.app_name)
```

**Comment by ccrvlh** [comment]
I'm not really using pydantic, but I also can't read .env file even using find_dotenv().
Funny thing is that when runny in VSCode debugger, everything works fine. It is as if the app factory was async, and it runs faster than find_dotenv() actually finds the file.

**Comment by rsip22** [comment]
It seems to me that the example for using .env files that can be found on the documentation does not fully work because it shows how to create the config file, which dependencies to install (`python-dotenv`), how to cache the settings... but it doesn't show that the function `get_settings` has to be called somewhere. Once I added `get_settings()` at some point on my main module, everything works and the values from `.env` file are correctly loaded. 

Perhaps adding an instruction about this to [the docs](https://fastapi.tiangolo.com/advanced/settings/#the-env-file) could be helpful?

**Comment by TotoMC-13** [comment]
@lowercase00 were you able to fix it? That if you use VSCode debugger it works fine, same thing happens to me.

**Comment by ccrvlh** [comment]
@TotoMC-13 I don't really remember, but I think you have to use both find_dotenv and load_dotenv (one finds the path, the other loads). But im not using Pydantic, so don't know how that would affect

**Comment by Xperhub** [comment]
I also got my .env file 

> In https://pypi.org/project/python-dotenv/ you can specifically call the path to your .env file:
> 
> ```
> ...
> # OR, explicitly providing path to '.env'
> from pathlib import Path  # Python 3.6+ only
> env_path = Path('.') / '.env'
> load_dotenv(dotenv_path=env_path)
> ```
> 
> I just got my .env to work by loading it in the main.py file before initiating the API.
> 
> ![2020-11-29 at 22 39](https://user-images.githubusercontent.com/2683346/100553033-e0523f80-3293-11eb-8946-17bfccd07d55.png)

This works for me. 

```
import os
from functools import lru_cache
from dotenv import load_dotenv



from pydantic import BaseSettings, Field
load_dotenv()
```

Then, I was able to call the instance of  my .env variables in my BaseSettings class

**Comment by khalmohak** [comment]
> I also got my .env file
> 
> > In https://pypi.org/project/python-dotenv/ you can specifically call the path to your .env file:
> > ```
> > ...
> > # OR, explicitly providing path to '.env'
> > from pathlib import Path  # Python 3.6+ only
> > env_path = Path('.') / '.env'
> > load_dotenv(dotenv_path=env_path)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I just got my .env to work by loading it in the main.py file before initiating the API.
> > ![2020-11-29 at 22 39](https://user-images.githubusercontent.com/2683346/100553033-e0523f80-3293-11eb-8946-17bfccd07d55.png)
> 
> This works for me.
> 
> ```
> import os
> from functools import lru_cache
> from dotenv import load_dotenv
> 
> 
> 
> from pydantic import BaseSettings, Field
> load_dotenv()
> ```
> 
> Then, I was able to call the instance of my .env variables in my BaseSettings class

Worked for me thanks

**Comment by yezz123** [comment]
You could use `python-decouple` just

```py
from decouple import config

SQLALCHEMY_DATABASE_URL = config("DATABASE_URL")
```

in the `.env`:

```sh
DATABASE_URL = "sqlite:///./database/dogeapi.sqlite"
```

**Comment by raindata5** [comment]
> dy try setting your `env_file` like this `env_file = "../.env"`

well this solved it for me @ycd thanks 😎👍!

**Comment by THOM-AwS** [comment]
I'm having this same issue here. I am using FastAPI, trying to import pydantic basesettings. still not working.

**Comment by ReeFish** [comment]
I also encountered the same problem, .env must be placed in the same path as main.py, I changed the path of .env, the problem is solved

**Comment by ReeFish** [comment]
In addition, there is a relative path method to solve, but the relative path is the relative path of the .env file relative to main.py, not the relative path relative to config.py (the Settings class can be in any path), just verified Yes, it works fine.
So there can be two solutions:
1) Put the .env file in the same path as main.py
2) Or set env_file as a relative path, the relative path of .env file relative to main.py
````
    class Config:
         env_file = '../.env'
         env_file_encoding = 'utf-8'
````

**Comment by ReeFish** [comment]
My directory structure is as follows:
```
├── api
│   ├── config.py
│   ├── main.py
├──-.env
```

**Comment by nikoBrajdic** [comment, marked as answer]
What the docs don't state obviously enough is that you **must declare the same fields in both the .env file AND the settings class!**
The mapping is _case-insensitive_.
Apart from this oversight I did everything else by the docs and it worked.
e.g.

.env
``` ini
USERNAME = foo
PaSsWoRd = bar
app_name = My App
```

settings.py
``` python
from pydantic import BaseSettings


class Settings(BaseSettings):
    app_name: str
    username: str
    password: str

    class Config:
        env_file: str = '.env'
```

main.py
``` python
from fastapi import FastAPI

from config.settings import Settings

app = FastAPI()
settings = Settings()

@app.get("/")
async def root():
    return {"settings": settings}

```

call from terminal
``` bash
$ curl localhost:8000
{"settings":{"app_name":"My App","username":"foo","password":"bar"}}
```

**Comment by jalvespinto** [comment]
I cannot get it to work either. By what says on the docs you don't have to declare any attributes on the `settings` class:
```
class Settings(BaseSettings):
    ...

    class Config:
        env_file = '.env'
        env_file_encoding = 'utf-8'
```

**Comment by jalvespinto** [comment]
I am thinking of using this hack until I can figure out how to do it the right way. Would I run into any issues doing this way? tks
```
class APPSetings(BaseSettings):
    def __init__(self, **kwargs):
        self.__dict__.update(kwargs)

@lru_cache()
def get_app_settings():
    env = os.getenv("APP_CONFIG", "development")
    env_path = f"{DOTENV_PATH}/.env.{env}"
    kwargs.update(dotenv_values(env_path))
    return config.APPSetings(**kwargs)
```

**Comment by plocher** [comment]
Thanks [nikoBrajdic](https://github.com/nikoBrajdic)

> What the docs don't state obviously enough is that you must declare the same fields in both the .env file AND the settings class!

Several tutorials out there suggest that any items not specified in the env file will be read from the OS environment:
>     # This is why you won’t see code like
>     #    SECRET_KEY: str = os.getenv("SECRET_KEY")
but don't bother to mention that waht you NEED to see is
>     #    SECRET_KEY: str


I still wish there was a "debug" mechanism to trace the initialization / .env processing stuff just to verify that it is working, since a badly named env_file doesn't throw any warnings...

**Comment by Insighttful** [comment]
Python-dotenv's internals contains a function that would be relevant to this discussion:

[find_dotenv](https://github.com/theskumar/python-dotenv/blob/137bc3dc0b8cf3d417a1e800c4065c526e3fb96a/src/dotenv/main.py#L270)
> Search in increasingly higher folders for the given file
> Returns path to the file if found, or an empty string otherwise

Since this module is what is being used for Pydantic's BaseSettings, it should be relatively simple use find_dotenv to locate the .env file automatically if in a parent directory, and save a lot of confusion. 

Maybe I am missing something, but why not:
```Python
from pydantic import BaseSettings
from dotenv import find_dotenv

class Settings(BaseSettings):
    host: str
    port: int
    user: str
    password: str

    class Config:
        env_file = find_dotenv()
        env_file_encoding = 'utf-8'
```

I wish this was default behavior without the Config class even being required.

**Comment by anthonycepeda** [comment]
You can use `pipenv` to run your app.
It will load the vars from your `.env` file before running the app:

```bash
$ pipenv run python main.py
Loading .env environment variables...
```

**Comment by TheHormat** [comment]
My solutions:
**.env file:**
```
from dotenv import load_dotenv
from pydantic_settings import BaseSettings

load_dotenv()

class Settings(BaseSettings):
    database_url: str

    class Config:
        env_file = ".env"

settings = Settings()

print("Database URL from Settings:", settings.database_url)
```
